### PR TITLE
Move encode_metadata out of CrateStore.

### PR DESCRIPTION
**Pull request by cjgillot**
`rustc_metadata` is already accessible by all client crates. It does not need to be called trough a trait object.

### CODE REVIEW

**Comment by bjorn3** [comment]
Can you move EncodedMetadata to rustc_metadata?

File: compiler/rustc_codegen_cranelift/scripts/filter_profile.rs [hunk 1]
```diff
@@ -96,7 +96,7 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
             stack = &stack[..index + REPORT_SYMBOL_NAMES.len()];
         }
 
-        const ENCODE_METADATA: &str = "rustc_middle::ty::context::TyCtxt::encode_metadata";
+        const ENCODE_METADATA: &str = "rustc_metadata::encode_metadata";
```
Comment (bjorn3) [inline compiler/rustc_codegen_cranelift/scripts/filter_profile.rs:99]:
```suggestion
        const ENCODE_METADATA: &str = "rustc_metadata::rmeta::encoder::encode_metadata";
```

**Comment by Mark-Simulacrum** [comment]
@bors try @rust-timer queue

Calls through a trait object *might* be saving us some codegen time and such (if things are not getting inlined, etc.). Seems unlikely, but worth checking, I think.

**Comment by oli-obk** [comment]
the change lgtm, but considering it's just a redesign, but regresses perf, should we really do this? Do you need this for some other PRs?

**Pull request merged**